### ISSUE TITLE
Multiple packages in own repository

### ISSUE THREAD

**Issue by ghost** [composer/composer#2588]
Hey,
i just wann know if it is possible to create a own repository for all my packages.
Or do i have to create multiple branches? which isnt best aproach, i think.

I want to be able to install those packages individually.

May you could point me to the right direction ;)

**Comment by xavier83ar** [comment]
It would be great to have this feature. In our case, we are a small software factory which have some packages that we reuse on many project, we use composer a lot, and this packages are on the same svn repository. With something like this, we could add only one repo to composer.json of all projects, and then install whatever package it needs, right now doing that require us to include one repo for each package.

Also, if we take github, it changed time ago to reflect this common case, and introduced the concept of organization, where all repos for different projects of the same organization are placed.

So, I think that this is something that composer should take into consideration. Having a way of adding a root svn or a git organization where to find packages. If there is a way of achieve this with other approach, I'd love to know  it, cause I couldn't find it.

**Comment by Seldaek** [comment]
@xavier83ar one option if you have many private packages in different repos is to use https://toranproxy.com and then you have just one (or two if you want packagist proxying) repo to add to the project and that gives you access to all your packages.

**Comment by Yehonal** [comment]
I've found a way to "bypass" this issue in this simple way:

remove the "master" branch from your repository , then having :

branch-first  ->  "name" : "first/directory"
branch-second  ->  "name" : "second/directory"

it works, installing it in different folders...

However @Seldaek you should consider to allow it, even if "usually it's not used" ... think about following situations:

1) a library that resides on vcs/git and after a specific version, it changes the **name** and so the path...having finally 2 branches: branch-ver1   , branch-ver2 ( with new name )

in this case i think you'll have a **bug** since the branch-ver1 still has "name": "name/old"   , instead branch-ver2 has the "name": "name/new"

but when someone try to install the old version , it will be installed in new path instead breaking different inclusions..


2) or maybe we just want different packages in same repository ( vcs) to avoid multiple repo because of they are strictly related to it. For example, in some personal repository i've:

- master ( for production ) -> "name": "myrepo/src" 
- test ( dev dependency )   -> "name": "myrepo/test" 
- doc  ( dev dependency )  -> "name": "myrepo/doc" 


Please consider to remove this useless limit just because it could be an edge case....it causes a lot of problems, moreover it can be bypassed

**Comment by FezVrasta** [comment]
Consider that some of the most popular JS projects make use of monorepos with Git...

React, Angular, Babel and many others.

**Comment by Johann-S** [comment]
Agreed with @FezVrasta a lot of populer project use Lerna to ship differents packages in one repository

**Comment by stof** [comment]
Well, composer was never meant to be a package manager for JS projects. It is logical that these projects optimize their workflow for npm, and may not satisfy the composer needs.

Npm packages can do what they want with their git repositories, because npm does not rely on the git repository at all. Library authors are submitting tarballs of code to the npm registry when publishing a release. And npm does not give you access to dev versions of packages like composer does (you can depend on a git repository directly, but this requires a special requirement syntax, and won't work for monorepos either).

**Comment by FezVrasta** [comment]
So are you suggesting Bootstrap to drop support for Composer because Composer isn't meant to serve JS projects, right?

Thanks for your time!

**Comment by stof** [comment]
My own thinking is indeed that if this prevents them from improving their own workflow, they may consider dropping support instead (if they don't switch to monorepo, they might keep it. But I'm sure I want use composer to install Bootstrap)

**Comment by stof** [comment]
Note that Bootstrap does not have composer support yet (the PR is not merged), so the question is not whether to drop it, but actually whether to add it (and then, I'm even more in favor of saying no)

**Comment by Johann-S** [comment]
We already have composer https://github.com/twbs/bootstrap/blob/v4-dev/composer.json but we want to ship several packages in one repository so IMO unfortunately we'll drop composer

**Comment by P0rnflake** [comment]
There are also single repos wich provide some of the JS libraries over composer, even for bootsrap (f.e. https://github.com/components/bootstrap) but they have always to copy the distribution and release it separately from the main project. If you can't add component support out of the box, maybe this is the way to go. For me managing the js/css/.. dependencies over composer as well is a benefit which i don't want to miss anymore, precicely because of the requirejs integration.

**Comment by filips123** [comment]
Support for multiple packages and packages in folders could also be used if some program has API clients in multiple languages, so each could have it's own package manager (NPM, PyPI, Composer) but in same repository.

For example:

* `libraries`: Folder with client libraries
* * `javascript`: Folder with JavaScript library
* * `java`: Folder with Java library
* * `python`: Folder with Python library
* * `php`: Folder with PHP library

**Comment by groovenectar** [comment]
It would be nice to see this functionality.

Already it works this way using the `path` repository type. I started to see value in this pattern, and working with it. I found out about the limitation while finally trying to move the system into remote VCS. It seems it's just not going to work this way using VCS, currently.

I'm wondering the feasibility of creating some kind of hybrid, so that the `VCS` type can scan in a similar way that `path` does, once it's cloned, perhaps with a glob as well...
